### PR TITLE
Implement first RL learning test

### PR DESCRIPTION
**Pull request by SchwammDev**
These automated tests using a most simple environment make sure that the REINFORCE algorithm is actually learning something and converging properly. They can also be used for fast diagnosis and understanding of how the mechanics of RL work. Because the environment is so simple, a
fast iteration cycle and step by step debugging is possible. Still it can shed light on many issues typically encountered when designing RL algorithm (i.e. normalization errors). It also gives a feel on what makes an algorithm stable and what not.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        "--plot-policy", action="store_true", default=False, help="plots policy parameter information for certain tests"
+    )
+
+
+class TrajectoryBuilder:
+    def __init__(self):
+        self._history = list()
+        self._current_record = None
+
+    def add(self, observation, action, reward):
```
Comment (SchwammDev) [inline tests/conftest.py]: Should be consistent with the rest: action, observation, reward

File: reinforcement/trajectories.py [hunk 2]
```diff
@@ -7,32 +7,78 @@ def __init__(self, observations, actions, returns):
         self.actions = actions
```
Comment (SchwammDev) [inline reinforcement/trajectories.py:7]: Switch order: actions, observations, returns for consistency

File: reinforcement/trajectories.py [hunk 3]
```diff
@@ -7,32 +7,78 @@ def __init__(self, observations, actions, returns):
...
+        self._observation = None
+        self._reward = None
 
-    def add(self, observation, action, reward):
-        self._history.append((observation, action, reward))
+    def start(self):
+        return self._ActionRecord(self)
+
+    def record_next(self):
+        self._history.append((self._observation, self._action, self._reward))
```
Comment (SchwammDev) [inline reinforcement/trajectories.py]: Consistency...

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+
+@pytest.fixture
+def make_trajectory(make_trajectory_builder):
+    def factory(actions=None, observations=None, returns=None):
+        t = make_trajectory_builder()
+        for a, o, r in zip(actions or itertools.repeat(0),
+                           observations or itertools.repeat(np.zeros((3, 2))),
+                           returns or range(0, 3)):
+            t.add(o, a, r)
```
Comment (SchwammDev) [inline tests/conftest.py]: Consistency

File: tests/test_agent_basis.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+def episode(batch_agent, algorithm, observation_factory, make_trajectory_builder):
+    class _Episode:
+        @staticmethod
+        def run():
+            b = make_trajectory_builder()
+            for r in range(-1, 2):
+                o = observation_factory()
+                a = batch_agent.next_action(o)
+                batch_agent.signal(r)
+                b.add(o, a, r)
```
Comment (SchwammDev) [inline tests/test_agent_basis.py]: consistency

File: tests/test_agent_basis.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    return _Episode()
+
+
+@pytest.mark.parametrize('invalid_agent_type', [MissingNextAction, MissingSignal, MissingTrain])
+def test_missing_agent_interface_throws(invalid_agent_type):
+    with pytest.raises(TypeError):
+        invalid_agent_type()
+
+
+def test_batch_agent_returns_action_as_one_hot_vector(batch_agent, observation):
```
Comment (SchwammDev) [inline tests/test_agent_basis.py]: It actually doesn't return as one_hot_vector but as scalar

File: tests/test_agent_basis.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    def do_transition():
+        a = batch_agent.next_action(observation)
+        batch_agent.signal(0)
+        return a
+
+    assert on_average(do_transition) == approx([0.8, 0.2], abs=0.05)
+
+
+def on_average(producer):
+    return np.mean([np.eye(1, 2, k=producer())[0] for _ in range(1000)], axis=0)
```
Comment (SchwammDev) [inline tests/test_agent_basis.py]: could use the np operations one_hot function

File: tests/test_rl_algorthims.py [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+
+
+class Space:
+    pass
+
+
+class Discrete(Space):
+    def __init__(self, n):
+        self.n = n
+        self.shape = (1,)
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: I don't think this is necessary

File: tests/test_rl_algorthims.py [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+        r = np.random.randint(1)
+        return np.array([r, 1 - r])
+
+
+class SimpleSwitchingMDP:
+    def __init__(self):
+        self.action_space = Discrete(2)
+        self.observation_space = BinaryBox()
+        self.len_episode = 100
+        self.max_total_return = 100
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: This is now a stochastic value, maybe the name should reflect that, and maybe it should be calculated from episode length and reward distributions

File: tests/test_rl_algorthims.py [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+        return 1 - observation[0]
+
+    def signal(self, reward):
+        pass
+
+    def train(self):
+        pass
+
+
+class NpSwitchingPolicy:
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: A better name for policy that reflects how it actually works. Maybe TanhPolicy or LinearPolicy

File: tests/test_rl_algorthims.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+            pass
+
+        def record(self, *args, **kwargs):
+            pass
+
+        def show(self):
+            pass
+
+
+class TfSwitchingPolicy:
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: Also better name, like ThetaPolicy, OneLayerPolicy or NN1Policy

File: tests/test_rl_algorthims.py [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+    def __init__(self, session, obs_dims, num_actions, make_sum_writer, lr=10):
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: make_sum_writer might be a misleading name, as it could also mean sum in the mathematical sense and not a short for summary

File: tests/test_rl_algorthims.py [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+        self._session = session
+        self._make_sum_writer = make_sum_writer
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: Same goes for the attribute

File: tests/test_rl_algorthims.py [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+
+        self._logs = [tf.compat.v1.summary.scalar("mean_normalized_return", tf.reduce_mean(self._in_returns)),
+                      self._log_2d_tensor_as_img("theta", theta)]
+        self._log_summary = tf.no_op
+        self._sum_writer = None
+        self._cur_iteration = 0
+
+    @staticmethod
+    def _log_2d_tensor_as_img(name, mat):
+        return tf.compat.v1.summary.image(name, tf.reshape(mat, shape=(1, mat.shape[0].value, mat.shape[1].value, 1)))
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: Maybe use the compat.v1 directly in the import statement to avoid these code duplications i.e. as tf1 or similar

File: tests/test_rl_algorthims.py [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+    return 100
+
+
+@pytest.fixture
+def eval_length():
+    return 100
+
+
+@pytest.fixture
+def simple_env():
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: better name; maybe switching_env?

File: tests/test_rl_algorthims.py [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+    return SimpleSwitchingMDP()
+
+
+@pytest.fixture
+def random_agent(simple_env):
+    return RandomAgent(simple_env.action_space)
+
+
+@pytest.fixture
+def dt_agent():
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: better name for this agent

File: tests/test_rl_algorthims.py [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+
+@pytest.fixture
+def session():
+    tf.compat.v1.reset_default_graph()
+    with tf.compat.v1.Session() as s:
+        yield s
+
+
+@pytest.fixture
+def make_sum_writer(session, log_tensorboard, request):
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: sum again misleading

File: tests/test_rl_algorthims.py [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+    return writer
+
+
+@pytest.fixture
+def baseline():
+    return FixedBaseline(0.0)
+
+
+@pytest.fixture
+def np_policy(simple_env, plot_policy):
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: again better name for this policy

File: tests/test_rl_algorthims.py [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+    p.plot()
+
+
+@pytest.fixture
+def np_reinforce(np_policy, baseline):
+    return Reinforce(np_policy, 0.0, baseline)
+
+
+@pytest.fixture
+def np_reinforce_agent(np_reinforce):
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: maybe find a better fixture setup the tries out the different policies with the agents. Currently, this code is duplicated. Would handle similar to the way I've done in uva...

File: tests/test_rl_algorthims.py [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+
+@pytest.mark.flaky(reruns=2, reruns_delay=3)
+def test_random_agent_only_achieves_random_expected_total_return(simple_env, random_agent):
+    avg_return = run_sessions_with(simple_env, random_agent, 1000)
+    assert avg_return == approx(0.0, abs=2)
+
+
+def run_sessions_with(env, agent, num_sessions):
+    avg_total_return = 0
+    for e in range(num_sessions):
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: replace e with _ because unused

File: tests/test_rl_algorthims.py [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+        while not done:
+            obs, r, done, _ = env.step(agent.next_action(obs))
+            agent.signal(r)
+            total_r += r
+        agent.train()
+        avg_total_return += total_r / num_sessions
+    return avg_total_return
+
+
+def test_deterministic_agent_achieves_max_return(simple_env, dt_agent, eval_length):
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: Might fail because of deterministic is the approximation forgiving enough?

File: tests/test_rl_algorthims.py [hunk 22]
```diff
@@ -0,0 +1,355 @@
...
+    avg_return = run_sessions_with(simple_env, np_reinforce_agent, eval_length)
+    assert avg_return == approx(simple_env.max_total_return, abs=5)
+
+
+@pytest.mark.flaky(reruns=2, reruns_delay=3)
+def test_tf_reinforce_agent_learns_near_optimal_solution(simple_env, tf_reinforce_agent, train_length,
+                                                         eval_length):
+    run_sessions_with(simple_env, tf_reinforce_agent, train_length)
+    avg_return = run_sessions_with(simple_env, tf_reinforce_agent, eval_length)
+    assert avg_return > simple_env.max_total_return * 0.9
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: for consistency also approx? Also this two tests could maybe be collapsed via parameterization and fixture setup mentioned above

File: tests/test_rl_algorthims.py [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+def test_deterministic_agent_achieves_max_return(simple_env, dt_agent, eval_length):
+    avg_return = run_sessions_with(simple_env, dt_agent, eval_length)
+    assert avg_return == approx(simple_env.max_total_return, rel=1)
+
+
+@pytest.mark.flaky(reruns=2, reruns_delay=3)
+def test_np_reinforce_agent_achieves_near_optimal_solution(simple_env, np_reinforce_agent, train_length, eval_length):
+    run_sessions_with(simple_env, np_reinforce_agent, train_length)
+    avg_return = run_sessions_with(simple_env, np_reinforce_agent, eval_length)
+    assert avg_return == approx(simple_env.max_total_return, abs=5)
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: Or this just > fraction of average return?

File: tests/test_agent_basis.py [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+
+def test_batch_agents_samples_its_algorithm_for_next_action(batch_agent, algorithm, observation):
+    algorithm.returns_sample_probability = [0.8, 0.2]
+
+    def do_transition():
+        a = batch_agent.next_action(observation)
+        batch_agent.signal(0)
+        return a
+
+    assert on_average(do_transition) == approx([0.8, 0.2], abs=0.05)
```
Comment (SchwammDev) [inline tests/test_agent_basis.py]: Could just check `returns_sample_probability`

File: reinforcement/models/neural_network.py [hunk 25]
```diff
@@ -51,7 +51,7 @@ def __init__(self, input_size, seed=None):
         :param input_size: size of the input vector
         :param seed: (optional) a seed used for initialization
         """
-        self._inputs = tf.placeholder(tf.float32, [None, input_size])
+        self._inputs = tf.compat.v1.placeholder(tf.float32, [None, input_size])
```
Comment (SchwammDev) [inline reinforcement/models/neural_network.py]: use tf1

File: reinforcement/algorithm/reinforce.py [hunk 26]
```diff
@@ -1,33 +1,34 @@
...
         self._gamma = gamma
         self._baseline = baseline
-        self._policy.set_signal_calc(_calc_signal)
+        self._policy.set_signal_calc(self._calc_signal)
+
+    @staticmethod
+    def _calc_signal(ops, in_actions, out_probs, in_returns):
+        one_hot = ops.one_hot(in_actions, out_probs.shape[1])
+        selected_prob = ops.reduce_sum(one_hot * out_probs, axis=1)
+        return -ops.reduce_mean(ops.log(selected_prob) * in_returns)
```
Comment (SchwammDev) [inline reinforcement/algorithm/reinforce.py]: maybe negate it on the agent side as it depends also on the model used

File: reinforcement/models/neural_network.py [hunk 27]
```diff
@@ -74,8 +74,8 @@ def add_layer(self, units, activation, weight_init, bias_init='zeros'):
         self._layer_idx += 1
 
     def _make_weights(self, init, units):
-        return tf.get_variable("W{}".format(self._layer_idx), shape=[self._last_units, units],
-                               initializer=self._initializer(init))
+        return tf.compat.v1.get_variable("W{}".format(self._layer_idx), shape=[self._last_units, units],
```
Comment (SchwammDev) [inline reinforcement/models/neural_network.py]: use tf1

File: reinforcement/trajectories.py [hunk 28]
```diff
@@ -2,37 +2,83 @@
...
+    def __str__(self):
+        return "\n".join(
+            f"pi -> {a} given: \n{o}\n R: {r}" for a, o, r in zip(self.actions, self.observations, self.returns))
+
+
+def history_to_trajectory(history):
+    if len(history) == 0:
+        raise TrajectoryError("Attempt to create trajectory from empty history.\nRecord some data first.")
+    ats, obs, rws = list(zip(*history))
+    history.clear()
```
Comment (SchwammDev) [inline reinforcement/trajectories.py]: This should probably done on the caller site not here

File: reinforcement/trajectories.py [hunk 29]
```diff
@@ -2,37 +2,83 @@
...
+    def start(self):
+        return self._ActionRecord(self)
+
+    def record_next(self):
+        self._history.append((self._action, self._observation, self._reward))
 
     def to_trajectory(self):
-        obs, ats, rws = list(zip(*self._history))
-        return Trajectory(np.array(obs), np.array(ats), np.array(rws, dtype=np.float32))
+        return history_to_trajectory(self._history)
```
Comment (SchwammDev) [inline reinforcement/trajectories.py]: Put the clear here

File: tests/conftest.py [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+class TrajectoryBuilder:
+    def __init__(self):
+        self._history = list()
+        self._current_record = None
+
+    def add(self, action, observation, reward):
+        self._history.append((action, observation, reward))
+
+    def to_trajectory(self):
+        return history_to_trajectory(self._history)
```
Comment (SchwammDev) [inline tests/conftest.py]: don't forget to clear

File: tests/test_rl_algorthims.py [hunk 31]
```diff
@@ -0,0 +1,359 @@
...
+    def _flip_state(self):
+        self._current_state = np.array([self._current_state[1], self._current_state[0]])
+
+    def _is_done(self):
+        return self._episode == self.len_episode
+
+    def __repr__(self):
+        return f"SimpleDeterministic(action_space={repr(self.action_space)}, " \
+               f"observation_space={repr(self.observation_space)}, len_episode={self.len_episode}," \
+               f"max_total_return={repr(self.avg_total_return)})"
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: string needs to be fixed

File: tests/test_rl_algorthims.py [hunk 32]
```diff
@@ -0,0 +1,359 @@
...
+    return avg_total_return
+
+
+def test_optimal_agent_achieves_max_return(switching_env, optimal_agent, eval_length):
+    avg_return = run_sessions_with(switching_env, optimal_agent, eval_length)
+    assert avg_return == approx(switching_env.avg_total_return, rel=1)
+
+
+@pytest.mark.flaky(reruns=2, reruns_delay=3)
+def test_tf_reinforce_agent_learns_near_optimal_solution(switching_env, reinforce_agent, train_length,
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: test name should reflect new generalized test

**Review by SchwammDev — COMMENTED** [review]
fix remaining

File: tests/test_rl_algorthims.py [hunk 33]
```diff
@@ -0,0 +1,355 @@
...
+
+
+class SwitchingMDP:
+    def __init__(self):
+        self.action_space = Discrete(2)
+        self.observation_space = BinaryBox()
+        self.len_episode = 100
+        self._reward = NormalDistribution(1, 0.1)
+        self._penalty = NormalDistribution(-1, 0.1)
+        self.avg_total_return = self.len_episode * self._reward.mean
```
Comment (SchwammDev) [inline tests/test_rl_algorthims.py]: Better name avg_max_reward

**Review by SchwammDev — COMMENTED** [review]
fix
